R/test-module.R

```diff
@@ -55,53 +56,58 @@
 #' }, !!multiplier_arg_name := 2, !!!more_args)
 #' @export
 testModule <- function(module, expr, ...) {
-  expr <- substitute(expr)
-  .testModule(module, expr, ...)
+  .testModule(
+    module,
+    quosure = rlang::enquo(expr),
+    dots = rlang::list2(...),
+    env = rlang::caller_env()
+  )
 }
 
 #' @noRd
 #' @importFrom withr with_options
-.testModule <- function(module, expr, ...) {
-  # Capture the environment from the module
-  # Inserts `session$env <- environment()` at the top of the function
+.testModule <- function(module, quosure, dots, env) {
+  # Modify the module function locally by inserting `session$env <-
+  # environment()` at the beginning of its body. The dynamic environment of the
+  # module function is saved so that it may be referenced after the module
+  # function has returned. The saved dynamic environment is the basis for the
+  # `data` argument of tidy_eval() when used below to evaluate `quosure`, the
+  # test code expression.
   body(module) <- rlang::expr({
-    session$env <- environment()
+    session$env <- base::environment()
     !!!body(module)
   })
 
-  # Create a mock session
   session <- MockShinySession$new()
+  on.exit(if (!session$isClosed()) session$close())
+  args <- append(dots, list(input = session$input, output = session$output, session = session))
 
-  # Parse the additional arguments
-  args <- rlang::list2(..., input = session$input, output = session$output, session = session)
-
-  # Initialize the module
   isolate(
     withReactiveDomain(
       session,
       withr::with_options(list(`shiny.allowoutputreads`=TRUE), {
-        # Remember that invoking this module implicitly assigns to `session$env`
-        # Also, assigning to `$returned` will cause a flush to happen automatically.
+        # Assigning to `$returned` causes a flush to happen automatically.
         session$returned <- do.call(module, args)
       })
     )
   )
 
-  # Run the test expression in a reactive context and in the module's environment.
-  # We don't need to flush before entering the loop because the first expr that we execute is `{`.
-  # So we'll already flush before we get to the good stuff.
+  # Evaluate `quosure` in a reactive context, and in the provided `env`, but
+  # with `env` masked by a shallow view of `session$env`, the environment that
+  # was saved when the module function was invoked. flush is not needed before
+  # entering the loop because the first expr executed is `{`.
   isolate({
     withReactiveDomain(
       session,
       withr::with_options(list(`shiny.allowoutputreads`=TRUE), {
-        eval(expr, new.env(parent=session$env))
+        rlang::eval_tidy(
+          quosure,
+          data = rlang::as_data_mask(as.list(session$env)),
```

Comment:
My only concern with the masking approach is that it is only viable when the mask is shallow and distinct from the lexical envs. This is the case here because (IIUC) the mask is a clone of the execution env and doesn't include any parent environment of the module.

Are there cases where a module is a closure, i.e. where it depends on variables defined in one of the parents of the module function? In that case, these closure modules won't work properly with this approach.

Reply:
I think that case would be unusual, but if necessary users could access enclosed variables in parent scopes with e.g. `get("foo", session$env)`